### PR TITLE
pma: fix file growth infinite loop

### LINKED ISSUE #205: PMA goes into infinite loop when flist has exactly one block and you allocate something of exactly that size
I'm getting an infinite loop in the PMA.  Here's how it happens:

- state->flist has exactly one block.
- It allocates something which is the exact size of this block (in my case the size is just 1 page).
- in _flist_record_alloc it hits the [last case](https://github.com/urbit/ares/blob/status/rust/ares_pma/c-src/btree.c#L520). head is state->flist since there's only node and head->next is NULL and it frees head and sets state->flist to NULL.
- Now the next time the PMA tries to find a free block in _bt_falloc it fails to find one and calls _flist_grow to add one.
- _flist_grow calls ftruncate to increase the file size and calls _flist_insert to add the new memory to the flist.
- _flist_insert does not check if *dst is NULL and so the while loop immediately exists and it doesn't add any block.
- _bt_falloc tries again to find a free block and fails, so it calls _flist_grow and _flist_insert ot add more space, which doesn't check if *dst is NULL, ...

I tried a couple fixes for this such as making _flist_insert check if *dst is NULL and if so just allocating a new block, or doing this inside _flist_grow. Both cause crashes so I don't know what the real solution is.

### PR DESCRIPTION
**Pull request by barter-simsum**
resolves #205

### CODE REVIEW

**Comment by barter-simsum** [comment]
* not quite merge ready. this can potentially break `bt_free`. Freelist insertion needs to be revised

**Comment by barter-simsum** [comment]
@eamsden @philipquirk I believe this is merge-ready now. The first test in btest.c has been disabled because it directly calls _bt_insert. When closing and syncing the btree, this causes errors because the ranges we've pretended to allocate (but haven't actually) aren't mmaped. Leaving disabled for now until I make test changes to override the sync procedure it calls (to avoid actually syncing).

File: rust/ares_pma/c-src/btree.c [hunk 1]
```diff
@@ -2742,7 +2763,11 @@ bt_free(BT_state *state, void *lo, void *hi)
   BT_kv kv = leaf->datk[childidx];
   vaof_t offset = looff - kv.va;
   lopg = kv.fo + offset;
-  hipg = lopg + (looff - hioff);
+  hipg = lopg + (hioff - looff);
+
+  /* insert null into btree */
+  _bt_insert(state, looff, hioff, 0);
```
Comment (barter-simsum) [inline rust/ares_pma/c-src/btree.c:2769]: @eamsden Note, I've confirmed that bt_free is properly referencing the leaf's dirty bitmap (i.e. checking that the data ranges are actually dirty).

**Comment by philipquirk** [comment]
This crashes for me with: 

ares(8146,0x1dce1dec0) malloc: *** error for object 0x600003198090: pointer being freed was not allocated

**Comment by barter-simsum** [comment]
> This crashes for me with:
> 
> ares(8146,0x1dce1dec0) malloc: *** error for object 0x600003198090: pointer being freed was not allocated
> 
> It crashes the first time I call pma_sync()

@philipquirk  Can you give more context for the crash? When running it as a serf of vere, I get a crash, but not this one. For instance, can ycargo build --features "stop_for_debug"ou at least print the results of `info stack` after attaching to the ares process with gdb (build with the `"stop_for_debug"` feature).

**Comment by philipquirk** [comment]
> > This crashes for me with:
> > ares(8146,0x1dce1dec0) malloc: *** error for object 0x600003198090: pointer being freed was not allocated
> > It crashes the first time I call pma_sync()
> 
> @philipquirk Can you give more context for the crash? When running it as a serf of vere, I get a crash, but not this one. For instance, can you at least print the results of `info stack` after attaching to the ares process with gdb (build with the `"stop_for_debug"` feature).

I just tried it again and it no longer crashes on launch for me. I'm not sure what's difference. I'm not seeing any crash at the moment.

**Comment by philipquirk** [comment]
> > > This crashes for me with:
> > > ares(8146,0x1dce1dec0) malloc: *** error for object 0x600003198090: pointer being freed was not allocated
> > > It crashes the first time I call pma_sync()
> > 
> > 
> > @philipquirk Can you give more context for the crash? When running it as a serf of vere, I get a crash, but not this one. For instance, can you at least print the results of `info stack` after attaching to the ares process with gdb (build with the `"stop_for_debug"` feature).
> 
> I just tried it again and it no longer crashes on launch for me. I'm not sure what's difference. I'm not seeing any crash at the moment.

I just saw it crash when calling save in the serf (so probably in save_to_pma but i didn't attach a debugger)

**Comment by barter-simsum** [comment]
@eamsden have confirmed this doesn't crash on latest commit from status branch. Ready for review

**Review by eamsden — APPROVED** [review]

**Pull request merged**